pkg/cli/cli.go

```diff
@@ -248,20 +335,90 @@ func (c cli) buildRootCmd() *cobra.Command {
 	return rootCmd
 }
 
+// filterPluginsByKeys matches plugins known to a cli to a set of plugin keys,
+// returning unambiguously matched plugins.
+// A match occurs if:
+// - Name and version are the same.
+// - Long or short name is the same, and only one version for that name exists.
+// If long or short name is the same, but multiple versions for that name exist,
+// don't guess which version to use and instead return an error.
+func filterPluginsByKeys(versionedPlugins []plugin.Base, cliPluginKeys map[string]string) ([]plugin.Base, error) {
+	if len(cliPluginKeys) == 0 {
+		return versionedPlugins, nil
+	}
+
+	// Find all potential mateches for this plugin's name.
+	definitelyMatch, maybeMatch := []plugin.Base{}, map[string][]plugin.Base{}
+	for pluginName, pluginVersion := range cliPluginKeys {
+		cliPluginKey := plugin.Key(pluginName, pluginVersion)
+		// Prevents duplicate versions with the same name from being added to
+		// maybeMatch.
+		hasDefinitely := false
+		for _, versionedPlugin := range versionedPlugins {
+			longName := versionedPlugin.Name()
+			shortName := plugin.GetShortName(longName)
+
+			switch {
+			// Exact match.
+			case pluginName == longName && pluginVersion == versionedPlugin.Version():
+				definitelyMatch = append(definitelyMatch, versionedPlugin)
+				hasDefinitely = true
+			// Is at least a name match, and the CLI version wasn't set or was set
+			// and matches.
+			case pluginName == longName || pluginName == shortName:
+				if !hasDefinitely && (pluginVersion == "" || pluginVersion == versionedPlugin.Version()) {
```

Comment:
As you are `break`ing `if hasDefinetely` after this `switch`, you can avoid checking for a salfe `hasDefinetely` here.

```suggestion
				if pluginVersion == "" || pluginVersion == versionedPlugin.Version() {
```

Reply:
I think this code can be refactored a bit more too, will update.